### PR TITLE
Remove similar streamlines

### PR DESCRIPTION
**Pull request by frheault**
Script to remove similar/redundant streamlines.
Part of the processing for models and RBX or for density bias correction for machine learning (@ppoulin91 )

https://drive.google.com/open?id=14Z5Qi_atP9iDIFTvRck6h19hFb9ZZLbV

### CODE REVIEW

File: scilpy/segment/models.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+import numpy as np
+from dipy.align.bundlemin import distance_matrix_mdf
+from dipy.tracking.streamline import set_number_of_points
+from dipy.segment.clustering import QuickBundles
+from dipy.segment.metric import AveragePointwiseEuclideanMetric
+
+
+def remove_similar_streamlines(streamlines, threshold=5, do_avg=False):
+    """ Remove similar streamlines, shuffle streamlines will impact the results.
```
Comment (jchoude) [inline scilpy/segment/models.py]: `shuffle` -> `shuffling`

File: scilpy/segment/models.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    if do_avg:
+        return avg_streamlines
+    else:
+        return streamlines
+
+
+def subsample_clusters(cluster_map, streamlines, threshold,
+                       min_cluster_size, average_streamlines=False):
+    """ Using a cluster map, remove similar streamlines from all clusters
+    independently using chunk of 1000 streamlines at the time to prevent
```
Comment (jchoude) [inline scilpy/segment/models.py]: `chunk` -> `a chunk` or `chunks`

File: scilpy/segment/models.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        Contains the list of indices per cluster.
+    Parameters
+    ----------
+    streamlines : list of numpy.ndarray
+        Input streamlines to remove duplicates from.
+    threshold : float
+        Distance threshold to consider two streamlines similar, in mm.
+    min_cluster_size : int
+        Minimal cluster size to be considered. Clusters with less streamlines
+        that the provided value will de discarted.
```
Comment (jchoude) [inline scilpy/segment/models.py]: `discarted` -> `discarded`

File: scripts/scil_remove_similar_streamlines.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            args.min_cluster_size = 1
+
+            # Once the streamlines reached a low enough amount, switch to
+            # single thread for full comparison
+            if len(streamlines) < 10000:
+                args.processes = 1
+            random.shuffle(streamlines)
+
+    # After convergence, we can simply save the output
+    new_sft = StatefulTractogram(streamlines, sft, Space.RASMM)
```
Comment (GuillaumeTh) [inline scripts/scil_remove_similar_streamlines.py]: @frheault Can you use from_sft() ?

**Comment by mdesco** [comment]
pip install -r requirements fails on scikit-learn.
pip install sklearn works.
sklearn==0.0
scikit-learn==0.22.2.post1

**Comment by AntoineTheb** [comment]
> pip install -r requirements fails on scikit-learn.
> pip install sklearn works.
> sklearn==0.0
> scikit-learn==0.22.2.post1

Same thing happened in #121
`sklearn==0.0` should be removed from `requirements.txt`
`scikit-learn` version should be changed to `0.22.1`

File: scripts/scil_remove_similar_streamlines.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+Remove very similar streamlines from a bundle.
+Uses clustering to speed up the process. Streamlines are considered as similar
+based on a MDF threshold within each cluster. Can be used with large bundles,
+but the clustering parameters will need to be adjusted.
+
+The algorithm still use a system of chunks to ensure the amount of comparison
```
Comment (mdesco) [inline scripts/scil_remove_similar_streamlines.py]: use -> uses

File: scripts/scil_remove_similar_streamlines.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+(n**2) does not grow out of control. To overcome limitations related to this
+use of chunks, multiple iterations must be done until a convergence threshold
+is acheived.
```
Comment (mdesco) [inline scripts/scil_remove_similar_streamlines.py]: acheived -> achieved

File: scripts/scil_remove_similar_streamlines.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+def _buildArgsParser():
+    p = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter,
+                                description=__doc__)
+
+    p.add_argument('in_bundle',
+                   help='Path of the input bundle.')
+
+    p.add_argument('min_distance', type=float,
+                   help='Distance threshold for 2 streamlines to be '
+                        'considered similar (mm).')
```
Comment (mdesco) [inline scripts/scil_remove_similar_streamlines.py]: Can you add a few used cases of typical distance threshold? I have no clue if 1mm or 5mm is the right kind of options. Maybe here or above in the help. 

Something like, for example, for a CST -> X mm, for a UF -> Y mm...

File: scripts/scil_remove_similar_streamlines.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+    pool = multiprocessing.Pool(args.processes)
+    timer = time()
+
+    logging.debug('Lauching subsampling on %s processes.', args.processes)
+    last_iteration = False
+    while True:
+        if len(streamlines) < 1000:
+            logging.warning('Subsampling less than 1000 streamlines is risky.')
+            break
```
Comment (mdesco) [inline scripts/scil_remove_similar_streamlines.py]: Can you make this robust to number of iterations? Should not say this if original bundle has more than 1000 streamlines

File: scripts/scil_remove_similar_streamlines.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+                        'the first approximation [%(default)s].')
+    p.add_argument('--min_cluster_size', type=int, default=5,
+                   help='Minimum cluster size for the first iteration '
+                        '[%(default)s].')
+    p.add_argument('--convergence', type=int, default=100,
+                   help='Streamlines count difference threshold to stop '
+                        're-running the algorithm [%(default)s].')
+    p.add_argument('--avg_similar', action='store_true',
+                   help='Average similar streamlines rather than removing them'
+                        '[%(default)s]. Requires a small min_distance.')
```
Comment (mdesco) [inline scripts/scil_remove_similar_streamlines.py]: Maybe tell us a bit more about what it's doing and the purpose. It allows some "smoothing"

File: scripts/scil_remove_similar_streamlines.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    cluster_map = qbx_and_merge(ArraySequence(streamlines),
+                                thresholds,
+                                nb_pts=20,
+                                verbose=False)
+
+    return subsample_clusters(cluster_map, streamlines, min_distance,
+                              min_cluster_size, average_streamlines)
+
+
+def _buildArgsParser():
```
Comment (arnaudbore) [inline scripts/scil_remove_similar_streamlines.py]: rename: _build_arg_parser

File: scripts/scil_remove_similar_streamlines.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+
+import argparse
+from itertools import repeat, chain
+import logging
+import multiprocessing
+import random
+from time import time
+
+from dipy.io.stateful_tractogram import Space, StatefulTractogram
```
Comment (arnaudbore) [inline scripts/scil_remove_similar_streamlines.py]: Space not used
